velox/functions/prestosql/aggregates/MaxSizeForStatsAggregate.cpp

```diff
@@ -0,0 +1,239 @@
+  }
+
+  void extractAccumulators(char** groups, int32_t numGroups, VectorPtr* result)
+      override {
+    // Partial and final aggregations are the same.
+    extractValues(groups, numGroups, result);
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::template updateGroups<true, int64_t>(
+        groups,
+        rows,
+        args[0],
+        [](int64_t& result, int64_t value) {
+          if (result < value) {
+            result = value;
+          }
+        },
+        mayPushdown);
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool mayPushdown) override {
+    BaseAggregate::updateOneGroup(
+        group,
+        rows,
+        args[0],
+        [](int64_t& result, int64_t value) {
+          result = std::max(result, value);
+        },
+        [](int64_t& result, int64_t value, int /* unused */) {
+          result = value;
+        },
+        mayPushdown,
+        (int64_t)0);
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdateSingleGroup(group, rows, args[0]);
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /*mayPushdown*/) override {
+    doUpdate(groups, rows, args[0]);
+  }
+
+ protected:
+  void upsertOneAccumulatorNew(
+      char* const group,
+      std::vector<vector_size_t>& rowSizes,
+      vector_size_t idx) {
+    if (decoded_.isNullAt(idx)) {
+      return;
+    }
+
+    // Clear null.
+    clearNull(group);
+    // Set max(current, this).
+    int64_t& current = *value<int64_t>(group);
+    current = std::max(current, (int64_t)rowSizes[idx]);
+  }
+
+  void
+  doUpdate(char** groups, const SelectivityVector& rows, const VectorPtr& arg) {
+    decoded_.decode(*arg, rows, true);
+
+    if (decoded_.isConstantMapping() && decoded_.isNullAt(0)) {
+      // There's nothing to do; all values are null.
+      return;
+    }
+
+    if (decoded_.isConstantMapping()) {
+      estimateSerializedSizes(arg, 1);
+      std::fill(elementSizes_.begin(), elementSizes_.end(), elementSizes_[0]);
+    } else {
+      estimateSerializedSizes(arg, arg->size());
+    }
+    rows.applyToSelected([&](vector_size_t i) {
+      upsertOneAccumulatorNew(groups[i], elementSizes_, i);
+    });
+  }
+
+  // Estimate the sizes of first numRows elements in vector.
+  void estimateSerializedSizes(VectorPtr vector, vector_size_t numRows) {
+    elementSizes_.resize(numRows);
+    std::fill(elementSizes_.begin(), elementSizes_.end(), 0);
+    elementSizePtrs_.resize(numRows);
+    elementIndices_.resize(numRows);
+
+    for (int i = 0; i < numRows; i++) {
```

Comment:
On repeated invocation, we only need to initialize values if 'resize' increases the number of elements and we only need to initialize new elements. Currently, we re-initialize on each invocation.

Reply:
I see, but what about garbage values left in the vector?

Reply:
There doesn't seem to be any garbage. For a given index, the values are always the same, no?